bigtable/bttest/grpc_error_injection.go

```diff
@@ -0,0 +1,307 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+/*
+This file provides interceptors to hook into grpc streaming calls
+and inject latency and errors into cbtemulator for testing
+*/
+package bttest
+
+import (
+	"fmt"
+	"log"
+	"math/rand"
+	"strconv"
+	"strings"
+	"time"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+var (
+	validStreamMethodSuffixes = []string{"ReadRows", "MutateRows"}
+)
+
+/*
+   Creates interceptors inject latency or errors into cbtemulator
+*/
+type EmulatorInterceptor struct {
+	LatencyTargets       latencyTargets
+	GrpcErrorCodeTargets grpcErrorCodeTargets
+}
+
+func (esib *EmulatorInterceptor) StreamInterceptor() grpc.ServerOption {
+	log.Println("Building Stream Server Interceptor:")
+	if len(esib.LatencyTargets) > 0 {
+		log.Printf(" - Latency Targets: %s\n", esib.LatencyTargets.String())
+	}
+	if len(esib.GrpcErrorCodeTargets) > 0 {
+		log.Printf(" - Error Targets: %s\n", esib.GrpcErrorCodeTargets.String())
+	}
+
+	return grpc.StreamInterceptor(esib.interceptorFunc())
+}
+
+func (esib *EmulatorInterceptor) interceptorFunc() grpc.StreamServerInterceptor {
+	return func(srv interface{},
+		ss grpc.ServerStream,
+		info *grpc.StreamServerInfo,
+		handler grpc.StreamHandler) error {
+
+		startTime := time.Now()
+
+		// Latency injection
+		pVal := rand.Float64() * 100
+		for _, lt := range esib.LatencyTargets {
+			if strings.HasSuffix(info.FullMethod, lt.methodSuffix) && pVal >= lt.percentile {
+				addedLatency := time.Until(startTime.Add(lt.expectedDuration))
+				if addedLatency > time.Duration(0) {
+					time.Sleep(addedLatency)
+				}
+			}
+		}
+
+		// Return any actual errors from handler
+		err := handler(srv, ss)
+		if err != nil {
+			return err
+		}
+
+		// If no actual errors, run error injection
+		eRand := rand.Float64() * 100
+		for _, gt := range esib.GrpcErrorCodeTargets {
+			if strings.HasSuffix(info.FullMethod, gt.methodSuffix) && eRand <= gt.stackedErrorRate {
+				return status.Error(gt.grpcErrorCode, "Injected Emulator Error")
+			}
+		}
+
+		return nil
+	}
+}
+
+/*
+   latencyTargets define how long it will take for a method
+   to return at different percentiles. e.g. ReadRows:p50:100ms
+*/
+
+type latencyTargets []latencyTarget
+
+// Set() and String() used by Flags lib
+func (lts *latencyTargets) Set(s string) error {
+	lt, err := newLatencyTarget(s)
+	if err != nil {
+		return err
+	}
+	*lts = append(*lts, *lt)
+	return nil
+}
+
+func (lts *latencyTargets) String() string {
+	var s []string
+	for _, v := range *lts {
+		s = append(s, v.String())
+	}
+	return strings.Join(s, ", ")
+}
+
+type latencyTarget struct {
+	methodSuffix     string
+	percentile       float64
+	expectedDuration time.Duration
+}
+
+// Create new latencyTarget from string like "ReadRows:p50:100ms"
```

Comment:
This seems not great to me as a way to read in and pass around data.

Why not have the latency targets specified in a json file? Then you can easily define unmarshaling into language types using json annotations and skip all the parsing logic.